packages/flutter_tools/lib/src/device.dart

```diff
@@ -159,40 +166,98 @@ abstract class DeviceManager {
       Future.wait<List<Device>?>(futureDevices),
     ]);
 
+    if (waitForDeviceToConnect == true) {
+      for (final DeviceDiscovery discoverer in _platformDiscoverers) {
+        discoverer.cancelWaitForDeviceToConnect();
+      }
+    }
+
     if (exactMatchCompleter.isCompleted) {
       return <Device>[await exactMatchCompleter.future];
     }
     return prefixMatches;
   }
 
-  /// Returns the list of connected devices, filtered by any user-specified device id.
-  Future<List<Device>> getDevices() {
+  /// Returns a list of devices filtered by the user-specified device
+  /// id/name (if applicable) and [filter].
+  ///
+  /// If [filter] is not provided, a default filter that requires devices to be
+  /// connected will be used.
+  ///
+  /// (iOS devices only) If [waitForDeviceToConnect] is true and an exact match
+  /// is found but the device is not connected, wait for the device to connect.
+  Future<List<Device>> getDevices({
+    DeviceDiscoveryFilter? filter,
+    bool waitForDeviceToConnect = false,
+  }) {
     final String? id = specifiedDeviceId;
+    filter ??= DeviceDiscoveryFilter();
     if (id == null) {
-      return getAllConnectedDevices();
+      return getAllDevices(filter: filter);
     }
-    return getDevicesById(id);
+    return getDevicesById(
+      id,
+      filter: filter,
+      waitForDeviceToConnect: waitForDeviceToConnect,
+    );
   }
 
   Iterable<DeviceDiscovery> get _platformDiscoverers {
     return deviceDiscoverers.where((DeviceDiscovery discoverer) => discoverer.supportsPlatform);
   }
 
-  /// Returns the list of all connected devices.
-  Future<List<Device>> getAllConnectedDevices() async {
+  /// Returns a list of devices filtered by [filter].
+  ///
+  /// If [filter] is not provided, a default filter that requires devices to be
+  /// connected will be used.
+  Future<List<Device>> getAllDevices({
+    DeviceDiscoveryFilter? filter,
+  }) async {
+    filter ??= DeviceDiscoveryFilter();
```

Comment:
nit, I think the `DeviceDiscoveryFilter` class can be made const, and then this can be made a default value:

```suggestion
    DeviceDiscoveryFilter filter = const DeviceDiscoveryFilter(),
  }) async {
```

Reply:
nvm, DeviceDiscoveryFilter() be made const